feat(product): product 패키지 생성 및 엔티티, 예외처리 정의

Ko-udon opened the pull request:
## 변경사항

<!-- 이 PR에서 어떤점들이 변경되었는지 기술해주세요. 가급적이면 as-is, to-be를 활용해서 작성해주세요.  -->
📌 AS-IS


📌 TO-BE
1. Product에 관한 패키지를 생성하였습니다.
2. Product 엔티티를 정의하였습니다.
3. Product 예외처리 정의하였습니다.
4. 예외처리 테스트를 진행하였습니다.
5. Product API 공동 작업을 위한 Controller, Service 클래스를 생성하였습니다.


## 체크 리스트 (Checklist)

pull request 전에 아래 체크 리스트들을 만족하는 지 확인한 후 체크('x') 표시를 해주시기 바랍니다.

- [x]  master 브랜치가 아니라 **develop** 브랜치에 Merge하도록 pull request를 작성 중이신가요?
- [x]  모든 **테스트**가 성공했나요?

ehddn committed:
feat(product): product 패키지 생성 및 엔티티, 예외처리 정의

1. Product에 관한 패키지를 생성
2. Product 엔티티 정의
3. Product 예외처리 정의
4. 예외처리 테스트 진행
5. Product API 공동 작업을 위한 Controller, Service 클래스 생성

seyoung5744 commented on src/main/java/com/zerobase/wishmarket/domain/product/model/Product.java:
```diff
@@ -0,0 +1,41 @@
+import org.hibernate.envers.AuditOverride;
+import org.hibernate.envers.Audited;
+
+@Entity
```
@Entity 는 구분 명확하게 클래스명 근처로 옮겨주는게 좋을거 같아요

Ko-udon replied:
네 그게 더 좋을 것 같습니다!

seyoung5744 commented on src/main/java/com/zerobase/wishmarket/domain/product/model/Product.java:
```diff
@@ -0,0 +1,41 @@
+
+@Entity
+@Getter
+@Setter
```
@Setter 로 열어두면 DB 에 데이터가 잘못 반영될 가능성이 있을 거 같다는 생각이 드는데
기능별로 따로 메소드를 작성해서 set을 하는게 어떨까요?

Ko-udon replied:
이부분은 제가 잘 몰랐습니다!
날짜 변경을 위해 필요할지도 모른다고 생각하여 넣었는데
AuditingEntityListener 를 사용하여 @CreatedDate, @LastModifiedDate 값을 자동으로 넣어주는걸 알게되었습니다. 감사합니다~

seyoung5744 commented on src/main/java/com/zerobase/wishmarket/domain/product/model/Product.java:
```diff
@@ -0,0 +1,41 @@
+
+    private String productImage;
+
+    private String category;
```
Entity에서는 catogory가 String이지만 
나중에 Enum으로 관리해서 전역적으로 사용하는 방법도 있을거 같아요.

```java
public enum Category {
    TOP("001"), // 상의 001
    OUTER("002"), // 아우터 002
    PANTS("003"), //  바지 003
    BAG("004"), // 가방 004
    SHOES("005"), // 신발  005
    CLOCK("006"), // 시계 006
    HAT("007"), // 모자 007
    SOCKS("008"), // 양말/레그웨어 008
    SUNGLASSES("009"); // 선글라스/안경테 009

```

Kim-SeonBeom replied:
추후에 category table을 추가할때 고려하겠습니다

seyoung5744 commented on src/main/java/com/zerobase/wishmarket/domain/product/model/Product.java:
```diff
@@ -0,0 +1,41 @@
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
+@Audited
```
@Audited 를 사용하는 이유가 따로 있을까요??

Kim-SeonBeom replied:
현재로서는 product entity에 이력관리를 할요소가 없다고 판단되어 수정하겠습니다
추후에 History가 필요한 테이블에 사용해보겠습니다

seyoung5744 commented on src/main/java/com/zerobase/wishmarket/domain/product/model/Product.java:
```diff
@@ -0,0 +1,41 @@
+@Setter
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
```
@NoargsConstructor(AccessLevel.PROTECTED) 기본 생성자의 접근 제어를 PROTECTED로 설정해놓게 되면 무분별한 객체 생성에 대해 한번 더 체크할 수 있는 수단이 되기 때문에 고려해보는 게 좋을 거 같아요

관련 링크 : https://cobbybb.tistory.com/14

ehddn committed:
refactor(product): product PR 내용 반영

1. Product -> ProductEntity 로 클래스네임 변경
2. @Entity 어노테이션을 해당 클래스 바로 위에 위치변경
3. ProductEntity 수정
- @setter 삭제
- @NoArgsContructor (access = AccessLevel.PROTECTED) 추가하여 불완전한 객체 생성 방지
